src/storage/engine/mod.rs

```diff
@@ -335,8 +336,12 @@ pub struct Cursor<I: Iterator> {
     max_key: Option<Vec<u8>>,
     valid: bool, // whether or not the cursor is hitting the `min_key` and `max_key` bound
 
-    cur_key_has_read: bool,
-    cur_value_has_read: bool,
+    // Use `UnsafeCell` to wrap these flags to be efficient and provides interior mutability.
+    // We don't care the case that there might be race conditions over these flags: we allow
+    // this kind of inaccuracy in statistics. Also we don't need to use atomic variables since
+    // there is extra cost for atomic variables.
+    cur_key_has_read: UnsafeCell<bool>,
```

Comment:
What about `RefCell`?

Reply:
It's a good and safe choice! But it will introduce the cost of borrow checking to ensure correctness, while in this case in fact we allow race conditions (it doesn't need to be very accurate for statistics), so we use UnsafeCell here. Are you satisfied with this explanation?

Reply:
Why not use `Cell`?

Reply:
@BusyJay You are genius!